multiverse-source/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+mod point;
+
+use anyhow::Result;
+use dcspark_blockchain_source::{EventObject, PullFrom, Source};
+use multiverse::{BestBlock, BestBlockSelectionRule, Variant};
+pub use point::Point;
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+};
+
+pub struct Multiverse<K, V, InnerSource> {
+    multiverse: multiverse::Multiverse<K, V>,
+    source: InnerSource,
+    confirmation_depth: usize,
+    confirmed: Option<K>,
+}
+
+impl<K, V, InnerSource> Multiverse<K, V, InnerSource> {
+    pub fn new(
+        multiverse: multiverse::Multiverse<K, V>,
+        confirmation_depth: usize,
+        inner_source: InnerSource,
+    ) -> Self
+    where
+        K: AsRef<[u8]> + Eq + Hash + Debug + Clone + Sync,
+        V: Variant<Key = K> + Clone,
+    {
+        let BestBlock {
+            selected,
+            discarded: _,
+        } = {
+            let _span =
+                tracing::span!(tracing::Level::INFO, "selecting best root options").entered();
+            multiverse.select_best_block(BestBlockSelectionRule::LongestChain {
+                depth: confirmation_depth,
+                // not going to delete anything here, so this doesn't matter
+                age_gap: 0,
+            })
+        };
+
+        Self {
+            multiverse,
+            confirmation_depth,
+            source: inner_source,
+            confirmed: selected.map(|k| k.inner().clone()),
+        }
+    }
+
+    pub fn into_inner(self) -> InnerSource {
+        self.source
+    }
+}
+
+#[async_trait::async_trait]
+impl<K, V, InnerSource, P> Source for Multiverse<K, V, InnerSource>
+where
+    InnerSource: Source<Event = V, From = Vec<P>> + Send,
+    P: Point<V = V> + PartialEq,
+    K: AsRef<[u8]> + Eq + Hash + Debug + Clone + Display + PullFrom + Sync,
```

Comment:
3 single-letter types are quite difficult to read. I think we should improve the names

Reply:
`K` and `V` are exactly the types of `K` and `V` in the `Multiverse` struct. I don't think it adds more clarity for those to be different, unless we change them in the `Multiverse` struct too. Although I could call them `MultiverseK` and `MultiverseV`. I don't mind calling those `Key` and `Value`, but I think consistency is better.

I'll think about `P`.

Reply:
We can keep K and V, but there should be more clear connection with P and better name for P as well